### PR TITLE
Avoid creating "steps" in deconstruction binding and optimize output

### PR DESCRIPTION
**Pull request by jcouv**
The initial binding for deconstruction now produces a `BoundAssignmentOperator` (instead of a `BoundDeconstructionAssignmentOperator`).

The `BoundAssignmentOperator` has:
 - a `BoundTupleLiteral` as its Left,
 - a `BoundConversion` as its Right, holding:
     - a tree of `Conversion` objects with Kind=Deconstruction, information about a Deconstruct method (optional) and an array of nested `Conversions` (like a tuple conversion),
     - an `BoundExpression` as its Operand.

A number of optimizations just fell out from this change. Most deconstructions don't construct a `ValueTuple` anymore and most un-necessary temps are now avoided.

Fixes https://github.com/dotnet/roslyn/issues/16869 (don't construct ValueTuple in simple deconstruction)
Fixes https://github.com/dotnet/roslyn/issues/13632 (don't use temps in simple deconstruction)
Fixes https://github.com/dotnet/roslyn/issues/13631 (don't construct ValueTuple in simple deconstruction)
Fixes https://github.com/dotnet/roslyn/issues/13270 (Phase ordering and bound node design issues for tuple deconstruction)
Fixes https://github.com/dotnet/roslyn/issues/16311 (Avoid creating the tuple return type for deconstruction twice)
Fixes https://github.com/dotnet/roslyn/issues/16166 (don't construct ValueTuple un-necessarily)

@dotnet/roslyn-compiler for review.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 1]
```diff
@@ -98,33 +98,30 @@ public void Deconstruct(out int a, out string b)
...
                 string V_1, //y
                 int V_2,
-                string V_3,
-                string V_4)
+                string V_3)
   IL_0000:  newobj     ""C..ctor()""
   IL_0005:  ldloca.s   V_2
   IL_0007:  ldloca.s   V_3
-  IL_0009:  call       ""void C.Deconstruct(out int, out string)""
+  IL_0009:  callvirt   ""void C.Deconstruct(out int, out string)""
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: I'm not sure why this changed from `call` to `callvirt`. I'll check with Vlad on whether this is significant.

Comment (VSadov) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: that is ok. Most likely reason is that Deconstruct is now called off a temporary.

I.E. you had

```C#
(new C1()).Deconstruct
```

now it is in, probably more conservatively correct, form

```C#
var ephemeralTemp = new C();
ephemeralTemp.Deconstruct
```

Optimizer kills ephemeralTemp, but the call still looks like applied to something that could be `null`

Not a big deal. If needed to be fixed, it should probably be fixed in the IL generator.


---
In reply to: [100636485](https://github.com/dotnet/roslyn/pull/17027#discussion_r100636485) [](ancestors = 100636485)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 2]
```diff
@@ -840,84 +829,21 @@ static void Main()
...
-  IL_007d:  box        ""long""
-  IL_0082:  ldstr      "" ""
-  IL_0087:  ldloc.0
-  IL_0088:  box        ""int""
-  IL_008d:  call       ""string string.Concat(object, object, object)""
-  IL_0092:  call       ""void System.Console.WriteLine(string)""
-  IL_0097:  ret
-}
-");
+  IL_0002:  ldc.i4.2
```
Comment (VSadov) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Yeah, this is definitely more optimal :-)

Now that we do not bounce the values off an intermediate tuple variable, optimizer can figure what is going on, elides dead assignments and then kills unused temps

**Comment by alrz** [comment]
Sorry to interrupt, but I feel #16166 could be also addressed here, maybe?

**Comment by VSadov** [comment]
#16166 is likely fixed by this, but needs a test.

Now that locals scheduler can understand and optimize the lowered form, I'd expect

int M()
{
  var (_, e) = (1, 2);
  return e;
}

Be emitted as

int M()
{
  return 2;
}

**Comment by alrz** [comment]
So it's orthogonal to the discard itself, as it would not emit locals if they were not used. Great!

**Comment by jcouv** [comment]
I'll check the IL for the discard scenario `(x, _) = (1, 2);` and review the comments. Thanks!

**Comment by jaredpar** [comment]
This change will optimize the new constructor initialization pattern correct?  

``` csharp
class  C { 
  int field1, field2;

  C(int p1, int p2) => (field1, field2) = (p1, p2);
}
```

If so can we get a test + the IL diffing to verify we don't regress this case?  Probably need a variant for fields, properties (get + set), properties (get only) and ref returning properties,

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 3]
```diff
@@ -5227,6 +5108,35 @@ static void Main()
...
+
+            var comp = CompileAndVerify(source, additionalRefs: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("C.M()", @"
+{
+  // Code size        8 (0x8)
+  .maxstack  1
+  .locals init (C V_0)
+  IL_0000:  newobj     ""C..ctor()""
+  IL_0005:  stloc.0
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: I'm not sure why the result of `new C()` is stored into a local here. I'll chat with Vlad.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 4]
```diff
@@ -6243,5 +6155,162 @@ public static void Main()
...
+            var comp = CompileAndVerify(source, expectedOutput: "1 hello 2", additionalRefs: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("C..ctor(int, string, ref int)", @"
+{
+  // Code size       39 (0x27)
+  .maxstack  4
+  .locals init (long V_0,
+                string V_1,
+                int V_2,
+                long V_3)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: There is room for further optimization. I don't think we need `V_1` or `V_3` here.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs [hunk 5]
```diff
@@ -5,248 +5,300 @@
...
-                        args.targets.Add(variable);
-                    }
-                    else
-                    {
-                        args.targets.Add(args.self.TransformCompoundAssignmentLHS(variable, args.stores, args.temps, isDynamicAssignment: variable.Type.IsDynamic()));
-                    }
-                },
-                (temps: temps, stores: stores, targets: assignmentTargets, self: this)
-            );
+            AddPlaceholderReplacement(deconstruction._inputPlaceholder, target);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: I was hoping this change would eliminate the need for the placeholders.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: (Because placeholders are a nonlocal mechanism in the rewriter)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs (hunk 3 above)
Comment (VSadov) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: You have a pattern : ` unused1 = unused2 = Expr`
Optimizer can see that `unused1` is never read from, but it is not as smart to figure that `unused2` is also unused. 

That should be ok for the purpose of this change. 

Many discards should not even make it to the codegen, but, IMO, optimizing discards specifically should be done separately.

---
In reply to: [100868940](https://github.com/dotnet/roslyn/pull/17027#discussion_r100868940) [](ancestors = 100868940)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs [hunk 6]
```diff
@@ -5,248 +5,300 @@
...
+                    case BoundKind.TupleLiteral:
+                        var tuple = (BoundTupleExpression)variable;
+                        assignmentTargets.Add(new Binder.DeconstructionVariable(GetAssignmentTargetsAndSideEffects(tuple, temps, effects), tuple.Syntax));
+                        break;
+
+                    case BoundKind.ConvertedTupleLiteral:
+                        throw ExceptionUtilities.UnexpectedValue(variable.Kind);
+
+                    default:
+                        var temp = this.TransformCompoundAssignmentLHS(variable, effects, temps, isDynamicAssignment: variable.Type.IsDynamic());
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: Nice.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs [hunk 7]
```diff
@@ -67,6 +67,18 @@ internal bool IsArrayIndex
             }
         }
 
+        private class DeconstructionUncommonData : UncommonData
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: If you are not using anything from UncommonData, perhaps making this a sibling of UncommonData derived from a common base would be more appropriate? - just to save on unnecessary fields.

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: nestedConversions might go to the common base though

---
In reply to: [100886606](https://github.com/dotnet/roslyn/pull/17027#discussion_r100886606) [](ancestors = 100886606)

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs [hunk 8]
```diff
@@ -213,6 +233,7 @@ internal static Conversion GetTrivialConversion(ConversionKind kind)
         internal static Conversion ImplicitDynamic => new Conversion(ConversionKind.ImplicitDynamic);
         internal static Conversion ExplicitDynamic => new Conversion(ConversionKind.ExplicitDynamic);
         internal static Conversion InterpolatedString => new Conversion(ConversionKind.InterpolatedString);
+        internal static Conversion Deconstruction => new Conversion(ConversionKind.Deconstruction);
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: These factories are for "trivial" conversions that makes sense when only have "kind". If Deconstruction always needs extra data, it should not have this factory.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs (hunk 5 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: It also seems easy to get into a case where a subtree is lowered or even evaluated more than once. In other similar cases we just pass temps as parameters.
Is  "deconstruction._invocation" too complex to be handled via some kind of `MakeDeconstructionInvocation(..., loweredOutputExpr, inputTemps)`  ?

---
In reply to: [100884679](https://github.com/dotnet/roslyn/pull/17027#discussion_r100884679) [](ancestors = 100884679)

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs [hunk 9]
```diff
@@ -869,5 +897,51 @@ public override int GetHashCode()
...
+    internal class DeconstructionInfo
+    {
+        internal DeconstructionInfo(BoundExpression invocation, BoundDeconstructValuePlaceholder inputPlaceholder,
+            ImmutableArray<BoundDeconstructValuePlaceholder> outputPlaceholders)
+        {
+             (_invocation, _inputPlaceholder, _outputPlaceholders) = (invocation, inputPlaceholder, outputPlaceholders);
+        }
+
+        internal BoundExpression _invocation;
+        internal BoundDeconstructValuePlaceholder _inputPlaceholder;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: Since we have performed the binding for invocation during initial binding anyways, it seems wasteful and also more work (to channel the information as a MethodSymbol and then reconstruct an invocation during lowering). We can discuss as a further refinement.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs [hunk 10]
```diff
@@ -5,248 +5,300 @@
...
         }
 
-        /// <summary>
-        /// Applies the deconstructions.
-        /// Adds any new locals to the temps and any new expressions to be evaluated to the stores.
-        /// </summary>
-        private void ApplyDeconstructions(BoundDeconstructionAssignmentOperator node, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> stores, ArrayBuilder<BoundValuePlaceholderBase> placeholders, BoundExpression loweredRight)
+        private BoundTupleLiteral ApplyDeconstructionConversion(ArrayBuilder<Binder.DeconstructionVariable> leftTargets,
+            BoundExpression right, Conversion conversion, ArrayBuilder<LocalSymbol> temps, DeconstructionSideEffects effects,
+            bool inInit)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: Any parts of the tree on the right which are above an invocation of `Deconstruct`.
So `(1, 2)` is in init. So is `e`.
But if `e` deconstructs into tuples or objects that can be further deconstructed, then any of the side-effects involved will go into the `effects.deconstructions` list, as opposed to the `effects.init` list.
I'll add a comment.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionKindExtensions.cs [hunk 11]
```diff
@@ -41,6 +41,7 @@ public static bool IsImplicitConversion(this ConversionKind conversionKind)
                 case ConversionKind.PointerToVoid:
                 case ConversionKind.NullToPointer:
                 case ConversionKind.InterpolatedString:
+                case ConversionKind.Deconstruction:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionKindExtensions.cs]: Yes, I ran into this exception:
```
        at Roslyn.Utilities.ExceptionUtilities.UnexpectedValue(Object o)
	at Microsoft.CodeAnalysis.CSharp.ConversionKindExtensions.IsImplicitConversion(ConversionKind conversionKind)
	at Microsoft.CodeAnalysis.CSharp.DiagnosticsPass.StripImplicitCasts(BoundExpression expr)
	at Microsoft.CodeAnalysis.CSharp.DiagnosticsPass.IsSameLocalOrField(BoundExpression expr1, BoundExpression expr2)
	at Microsoft.CodeAnalysis.CSharp.DiagnosticsPass.CheckForAssignmentToSelf(BoundAssignmentOperator node)
	at Microsoft.CodeAnalysis.CSharp.DiagnosticsPass.VisitAssignmentOperator(BoundAssignmentOperator node)
        ...
```

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs (hunk 8 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: I use this one for error cases. Any time `MakeDeconstructionConversion` runs into errors, it will return this trivial conversion instance. Maybe I should rename it to `DeconstructionFailed`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs [hunk 12]
```diff
@@ -1646,6 +1646,13 @@ private bool RegularPropertyAccess(BoundExpression expr)
 
         public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
         {
+            if (node.Left.Kind == BoundKind.TupleLiteral)
+            {
+                ((BoundTupleExpression)node.Left).VisitAllElements((x, self) => self.VisitLvalue(x), this);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: Thanks! Both suggestions (Assign and VisitLValue) work nicely.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs (hunk 7 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: I may keep that for a follow-up. My main goal was to not clutter UncommonData, but I didn't want to engage in any deep refactorings there.

File: src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs [hunk 13]
```diff
@@ -12,7 +13,13 @@ namespace Microsoft.CodeAnalysis.CSharp
...
     /// <summary>
     /// This portion of the binder converts deconstruction-assignment syntax (AssignmentExpressionSyntax nodes with the left
-    /// being a tuple expression or declaration expression) into a BoundDeconstructionAssignmentOperator (or bad node).
+    /// being a tuple expression or declaration expression) into a BoundAssignmentOperator (or bad node).
+    /// The BoundAssignmentOperator will have:
+    /// - a BoundTupleLiteral as its Left,
+    /// - a BoundConversion as its Right, holding:
+    ///     - a tree of Conversion objects with Kind=Deconstruction, information about a Deconstruct method (optional) and
+    ///         an array of nested Conversions (like a tuple conversion),
+    ///     - an BoundExpression as its Operand.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs]: >an [](start = 14, length = 2)

A typo "an" vs. "a"?

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs [hunk 14]
```diff
@@ -67,6 +67,18 @@ internal bool IsArrayIndex
...
+        private class DeconstructionUncommonData : UncommonData
+        {
+            internal DeconstructionUncommonData(DeconstructionInfo deconstructionInfoOpt, ImmutableArray<Conversion> nestedConversions)
+                : base(false, false, default(UserDefinedConversionResult), null, nestedConversions)
+            {
+                Debug.Assert(!nestedConversions.IsDefaultOrEmpty);
+                _deconstructionInfo = deconstructionInfoOpt;
+            }
+
+            internal DeconstructionInfo _deconstructionInfo;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: >_deconstructionInfo [](start = 40, length = 19)

readonly? Also, consider using name DeconstructionInfo if we intend to access it from outside.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/NoPia.cs [hunk 15]
```diff
@@ -1506,18 +1506,18 @@ public void TestDeconstruction()
 
             var expectedDiagnostics = new[]
             {
-                // (8,12): error CS1768: Type 'ValueTuple<T1, T2>' cannot be embedded because it has a generic argument. Consider setting the 'Embed Interop Types' property to false.
+               // (8,12): error CS1768: Type 'ValueTuple<T1, T2>' cannot be embedded because it has a generic argument. Consider setting the 'Embed Interop Types' property to false.
                 //     public (int, int) TestTuple()
-                Diagnostic(ErrorCode.ERR_GenericsUsedInNoPIAType, "(int, int)").WithArguments("System.ValueTuple<T1, T2>").WithLocation(8, 12),
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/NoPia.cs]: >.WithLocation(8, 12), [](start = 122, length = 21)

Why are we dropping location here?

File: src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs [hunk 16]
```diff
@@ -1327,28 +1325,16 @@ static void Main()
             comp.VerifyDiagnostics(
                 // (6,13): error CS8059: Feature 'tuples' is not available in C# 6.  Please use language version 7 or greater.
                 //         var (x1, x2) = Pair.Create(1, 2);
-                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion6, "(x1, x2)").WithArguments("tuples", "7").WithLocation(6, 13),
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs]: >.WithLocation(6, 13) [](start = 112, length = 20)

Why are we dropping locations? Please restore them here and in other similar places.

**Comment by AlekseyTs** [comment]
As we talked offline, given that we are reusing existing bound node and how we are doing this, we should ensure that behavior around the following is preserved:
- IOperation node corresponding to deconstruction and its content
- Result of SemanticModel.GetConversion on the deconstruction target
- Result of SemanticModel.GetTypInfo on the deconstruction target

File: src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs (hunk 16 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs]: I think there is a bug with the diffing logic when the test fails. The locations are only printed out correctly when no diagnostics are expected.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs [hunk 17]
```diff
@@ -869,5 +898,51 @@ public override int GetHashCode()
...
+        internal DeconstructionInfo(BoundExpression invocation, BoundDeconstructValuePlaceholder inputPlaceholder,
+            ImmutableArray<BoundDeconstructValuePlaceholder> outputPlaceholders)
+        {
+             (_invocation, _inputPlaceholder, _outputPlaceholders) = (invocation, inputPlaceholder, outputPlaceholders);
+        }
+
+        internal BoundExpression _invocation;
+        internal BoundDeconstructValuePlaceholder _inputPlaceholder;
+        internal ImmutableArray<BoundDeconstructValuePlaceholder> _outputPlaceholders;
     }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: Can the fields be `readonly`? And internal field names should start with uppercase character.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs [hunk 18]
```diff
@@ -869,5 +898,51 @@ public override int GetHashCode()
...
+                }
+
+                return new TreeDumperNode("conversion", self.Kind, sub);
+            }
+        }
+#endif
+    }
+
+    /// <summary>Stores all the information from binding for calling a Deconstruct method.</summary>
+    internal class DeconstructionInfo
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: Could this be a `struct`?

File: src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs [hunk 19]
```diff
@@ -164,7 +164,7 @@ internal virtual string Dump()
         }
 #endif
 
-        private string GetDebuggerDisplay()
+        internal string GetDebuggerDisplay()
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: Is this change needed?

Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: Yes, I added a `GetDebuggerDisplay` on `DeconstructionVariable` which calls this. It was very convenient while debugging.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs (hunk 18 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: Sure. I thought about it, but wasn't decisive. I'll change.

**Comment by jcouv** [comment]
@AlekseyTs I talked to Gen and wrote a test for `IOperation` to try it out.
A deconstruction is represented as a tree with an assignment operation and two `NoneOperation`s. That seems correct given that tuples don't implement `IOperation` yet.
The test cannot complete a verification because of that. That is the same behavior on any expression with a tuple, such as `x = (1, 2);`.
In short, I think that this PR doesn't worsen `IOperation`.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IOperationTests.cs [hunk 20]
```diff
@@ -55,5 +55,39 @@ static void Test2(int y, params int[] x)
...
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib(text, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.RegularWithIOperationFeature);
+            compilation.VerifyDiagnostics();
+            var tree = compilation.SyntaxTrees.Single();
+            var model = compilation.GetSemanticModel(tree);
+
+            var assignments = tree.GetRoot().DescendantNodes().OfType<AssignmentExpressionSyntax>().ToArray();
+            Assert.Equal("(x, y, z) = (1, 2, 3)", assignments[0].ToString());
+            Assert.Equal(null, model.GetOperation(assignments[0]));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IOperationTests.cs]: >Assert.Equal [](start = 12, length = 12)

Assert.Null?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IOperationTests.cs]: >model.GetOperation(assignments[0]) [](start = 31, length = 34)

I am curious what happens when we get an IOperation node for the containing statement instead? Can we reach to the node for deconstruction by looking at its children. We better not get a node implementing IAssignmentExpression this way.

Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IOperationTests.cs]: I didn't think about that. Thanks.
I can think of two mitigations: let the assignment operation be returned but neuter is (Kind = None, etc), or go back to a using a new bound node.
I've made a commit with the first option. The trouble is that an interface/type check still reveals it is an assignment :-(
We can discuss tomorrow and decide whether that is sufficient.

**Comment by jcouv** [comment]
FYI @gafter @VSadov The implementation with BoundAssignmentOperator made the deconstruction visible as an assignment IOperation. Given that we don't know that will be the correct IOperation implementation, I'm going back to a deconstruction-specific bound node (BoundDeconstructionAssignmentOperator).
This avoids accidentally committing to an observable IOperation design.

**Comment by gafter** [comment]
@jcouv Please don't do that. First, there are mitigations if we decide to expose it as something other than an assignment (i.e. return a different IOperation object for one vs the other). Second, we can change it later if that's what we want to do.

**Comment by jcouv** [comment]
@gafter In the 2.0 implementation and for other C# 7 nodes, `GetOperation` returns `null` (because the new bound nodes don't implement `IOperation`).
If we re-use `BoundAssignmentOperator` for deconstruction, then it is possible to get a non-null operation object for the deconstruction. For instance, by accessing the parent operation and asking it for its child (that's a hole we can't easily paper over).

That object has a `Kind` (which I agree we could fake to `None`), but it also implements `IAssignmentExpression` (which has `Target` and `Value`, see [here](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/Expression.cs,f3446277ba75976f)). We can't paper over type checks either.
When it comes time to implement IOperations for deconstruction, are we necessarily going to decide to re-use `IAssignmentExpression`? Mostly likely it will be insufficient to represent the information we need for a deconstruction.

So it seems that using a distinct bound node (which we could change later) is less binding and keeps our options open.

What do you see as the advantages of re-using the assignment bound node? In my mind it is a bit simpler and a bit more elegant, but it's hard to see those outweigh the issue with IOperation APIs (which some analyzers apparently already consume).

**Comment by gafter** [comment]
@jcouv Anything we do in anticipation of possible design strategies for `IOperation` here is wasted work. Whatever you're thinking of possibly doing proactively now will either be work that we could have done later, or work that we won't need at all.

**Comment by jcouv** [comment]
I may need to make a correction. I need to confirm whether the RTM behavior is to return `null` or rather an instance which implements no `IOperation` interface. I'll get back to you after I re-check.

**Comment by jcouv** [comment]
After re-running the IOperation test in a separate branch, I can confirm I made a mistake in the description above when I spoke of a `null` return value. The RTM version returns an object which implements `IOperation` but no other interface.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 21]
```diff
@@ -392,29 +392,8 @@
...
-  <Node Name="BoundDeconstructionAssignmentStep" Base="BoundNode">
-    <Field Name="Assignment" Type="BoundAssignmentOperator" Null="disallow"/>
-    <Field Name="OutputPlaceholder" Type="BoundDeconstructValuePlaceholder" Null="disallow"/>
-  </Node>
-
-  <Node Name="BoundDeconstructionConstructionStep" Base="BoundNode">
-    <Field Name="Construct" Type="BoundExpression" Null="disallow"/>
-    <Field Name="OutputPlaceholder" Type="BoundDeconstructValuePlaceholder" Null="disallow"/>
+    <Field Name="Left" Type="BoundTupleExpression" Null="disallow"/>
+    <Field Name="Right" Type="BoundExpression" Null="disallow"/>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: >BoundExpression [](start = 30, length = 15)

BoundConversion?

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IOperationTests.cs [hunk 22]
```diff
@@ -55,5 +56,45 @@ static void Test2(int y, params int[] x)
...
+            var compilation = CreateCompilationWithMscorlib(text, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.RegularWithIOperationFeature);
+            compilation.VerifyDiagnostics();
+            var tree = compilation.SyntaxTrees.Single();
+            var model = compilation.GetSemanticModel(tree);
+
+            var assignments = tree.GetRoot().DescendantNodes().OfType<AssignmentExpressionSyntax>().ToArray();
+            Assert.Equal("(x, y, z) = (1, 2, 3)", assignments[0].ToString());
+            IOperation operation1 = model.GetOperation(assignments[0]);
+            Assert.NotNull(operation1);
+            Assert.False(operation1 is IAssignmentExpression);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IOperationTests.cs]: >Assert.False(operation1 is IAssignmentExpression); [](start = 12, length = 50)

I think we should also assert the kind. Here and below.

**Comment by jcouv** [comment]
From discussion with Neal, it's ok to keep the use a new bound node, for now. I'll leave BoundDeconstructionAssignmentOperator.

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest  please

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

//C:\Users\dotnet-bot\.nuget\packages\Microsoft.VSSDK.BuildTools\15.0.26124-RC3\tools\VSSDK\Microsoft.VsSDK.targets(649,5): error MSB4018: The "FindInstalledExtension" task failed unexpectedly. [D:\j\workspace\windows_debug---0c191de0\src\Samples\VisualBasic\ImplementNotifyPropertyChanged\Impl\ImplementNotifyPropertyChangedVB.vbproj]

**Comment by jcouv** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs [hunk 23]
```diff
@@ -5,248 +5,317 @@
...
+        /// - the initialization phase includes side-effects from the left, followed by evaluations of the right
+        /// - the deconstruction phase includes all the invocations of Deconstruct methods and tuple element accesses below a Deconstruct call
+        /// - the conversion phase
+        /// - the assignment phase
+        /// </summary>
+        private BoundExpression RewriteDeconstruction(BoundTupleExpression left, Conversion conversion, BoundExpression right)
 
-            RemovePlaceholderReplacements(placeholders);
-            placeholders.Free();
+        {
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: >{ [](start = 8, length = 1)

There is an unnecessary empty line before this one

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs]: I'll fix that in my next PR. Thanks

**Comment by Logerfo** [comment]
Couldn't [this case](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABADAAmwEYBuAWAChsBmQgJnwGF8BvS/Dw24Aex4Bt8AWQAUASlbtO07AHZ8IgB4h8MBAFc4SfHTEA6ReQrSAvpRNA=) be optimized as well?

**Comment by jcouv** [comment]
@Logerfo It could, although to be honest I wouldn't prioritize it.
